### PR TITLE
Allow the use of Distribution in Representations and SkyCoord

### PR DESCRIPTION
**Pull request by mhvk**
This pull request is updates the `Distribution` class to allow it to be used inside representations. After the PRs to get things to work with `EarthLocation`, the changes here are rather minimal: support `concatenate` and a small fix in how `broadcast_arrays` is dealt with (plus a very small change in registration of representations, just for the tests to work more easily); the bulk of this PR is extra tests.

Note that the support is not yet complete. E.g., the way representations do `__repr__` means that the `repr` gives no sign that distributions are used. I think the best way to fix this, though, is through support of `array2string` for representations, which is a bit beyond what I can do here (similarly, `Distribution.__repr__` should go via `array2string` so that one can indicate what are the samples more easily). There perhaps also is a more general question of whether things like `SkyCoord` should just allow `Distribution` to be used, or whether they should also gain a `.distribution` property.

Anyway, I think it makes sense to proceed in small steps, so just go with what is here first... 

- [X] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

File: astropy/uncertainty/tests/test_containers.py
Comment (nstarman) [inline astropy/uncertainty/tests/test_containers.py:210]: Should there be a set of tests for CoordinateFrame?

**Review by nstarman — COMMENTED** [review]
It's wonderful how few changes are actually required!

File: astropy/coordinates/representation/base.py
Comment (mhvk) [reply astropy/coordinates/representation/base.py:609]: Done.

File: astropy/coordinates/representation/base.py
Comment (mhvk) [reply astropy/coordinates/representation/base.py:610]: Agreed that this is not a very nice heuristic... But for another time!

File: astropy/uncertainty/core.py
Comment (mhvk) [reply astropy/uncertainty/core.py:274]: Not really: if `result is NotImplemented`, it gets returned here too if `out is True`

File: astropy/uncertainty/core.py
Comment (mhvk) [reply astropy/uncertainty/core.py:276]: I think once we have more functions there will be an `elif function in DISPATCHED_FUNCTIONS` clause` -- I just copied this from `Masked`, but removed what was not yet needed...

File: astropy/uncertainty/function_helpers.py
Comment (mhvk) [reply astropy/uncertainty/function_helpers.py:75]: Yes, sadly. I guess the alternative would be to import `FUNCTION_HELPERS`, etc., inside `Distribution.__array_function__`, but then every function call is slowed down.

File: astropy/uncertainty/function_helpers.py
Comment (mhvk) [reply astropy/uncertainty/function_helpers.py:80]: I've instead added a docstring - it may well be that another override will need to check that `n_samples` is the same for all `Distribution` instances.

File: astropy/uncertainty/function_helpers.py
Comment (mhvk) [reply astropy/uncertainty/function_helpers.py:126]: The tuple is only constructed here, and I think it is clearer to be explicit here that we're adding two tuples.

File: astropy/uncertainty/function_helpers.py
Comment (mhvk) [reply astropy/uncertainty/function_helpers.py:128]: Yes, makes sense.

File: astropy/uncertainty/function_helpers.py
Comment (mhvk) [reply astropy/uncertainty/function_helpers.py:135]: I chose to use the `dict` because then it looks more like giving arguments to a function (which is happening here).

File: astropy/uncertainty/tests/test_containers.py
Comment (mhvk) [reply astropy/uncertainty/tests/test_containers.py:147]: yes, done.

File: astropy/uncertainty/tests/test_containers.py
Comment (mhvk) [reply astropy/uncertainty/tests/test_containers.py:188]: Not quite sure what you mean!

File: astropy/uncertainty/tests/test_containers.py
Comment (mhvk) [reply astropy/uncertainty/tests/test_containers.py:210]: The tests for `SkyCoord` are really limited - one should really add, say, a distribution in temperature for `AltAz`. If you prefer, I can remove the `SkyCoord` tests so we just focus on the representations (I kept it in a separate commit partially for that reason; I was just very happy it worked at all!!).

File: astropy/uncertainty/function_helpers.py
Comment (nstarman) [reply astropy/uncertainty/function_helpers.py:75]: Or we define a Protocol / ABC in a third file. No circular imports or slow-downs.

File: astropy/uncertainty/function_helpers.py
Comment (nstarman) [reply astropy/uncertainty/function_helpers.py:142]: Ok. Might require a noqa in the future. I think pyupgrade is the tool that "fixes" these.

File: astropy/uncertainty/tests/test_containers.py
Comment (nstarman) [reply astropy/uncertainty/tests/test_containers.py:210]: I'm definitely happy it works as well. I was just saying that the tests should probably mimic the object hierarchy and have CoordinateFrame tests between Representation and SkyCoord tests.

File: astropy/uncertainty/tests/test_containers.py
Comment (nstarman) [reply astropy/uncertainty/tests/test_containers.py:188]: Just that this test looks similar to its namesake in `super`, adding tests for the velocity. So if you call `super` in this test then...

Comment (mhvk) [reply astropy/uncertainty/tests/test_containers.py:188]: Ah, yes, I see. But tricky to use super here, since I need to pass on `CartesianDifferential` in `represent_as`.

File: astropy/uncertainty/core.py
Comment (nstarman) [reply astropy/uncertainty/core.py:276]: Fair. And given the scope I'll just resolve this.

File: astropy/uncertainty/function_helpers.py
Comment (nstarman) [reply astropy/uncertainty/function_helpers.py:133]: Ok. The slowdown is negligible.

File: astropy/uncertainty/tests/test_containers.py
Comment (mhvk) [reply astropy/uncertainty/tests/test_containers.py:210]: Agreed in principle,  but I think adding another skeleton test does not help that much. And full tests should be a different PR. 

Probably this all argues I should just remove the commit... Though it is sort-of nice to have at least a minimal guard against messing up SkyCoord...

Comment (nstarman) [reply astropy/uncertainty/tests/test_containers.py:210]: It's fine to keep.

File: astropy/uncertainty/function_helpers.py
Comment (nstarman) [reply astropy/uncertainty/function_helpers.py:75]: Any opinion on adding a `DistributionBase` or `AbstractDistribution` Protocol? e.g. in a new file `base.py`:

```python
from typing import runtime_checkable, Protocol

class AbstractDistribution(Protocol):
    @property
    def distribution(self): ...

    @property
    def dtype(self): ...

    @property
    def n_samples(self): ...
```

Comment (nstarman) [reply astropy/uncertainty/function_helpers.py:75]: Then we can check for `Distribution`-like stuff in an implementation-free manner and avoid circular imports.

Comment (mhvk) [reply astropy/uncertainty/function_helpers.py:75]: Something like that may be good, though I'd prefer to first revisit your old PR of merging how `Distribution` and `Masked` generate classes, etc. While doing the work on `Distribution`, I realized there are quite a few bits that are not all that logical, and are done better for `Masked`. But of course the problem is to find the time...

Comment (nstarman) [reply astropy/uncertainty/function_helpers.py:75]: Yeah, the issue I had with this abstract class generation technique was making it so that the wrapper classes could stack in an intuitive and introspectable way so that we could have a clean MRO and way to unwrap everything.

Comment (nstarman) [reply astropy/uncertainty/function_helpers.py:75]: So defer this to a future PR?

Comment (mhvk) [reply astropy/uncertainty/function_helpers.py:75]: Yes, let's. It would be nice to solve this for all classes that use `__array_function__` (`Distribution`, `Masked`, `Quantity`).

**Review by nstarman — APPROVED** [review]
LGTM! I like the code / test ratio.
We've identified some things for followup, so this PR is ready.
Probably want other Coordinates maintainers to OK this?

**Comment by mhvk** [comment]
Thanks! I asked @eteq offline for a quick review, as the other `uncertainty` maintainer. Though I think you've covered it, thanks!!

**Comment by pllim** [comment]
@mhvk , I'll leave it to your discretion on when to merge then. But I would advise against waiting till the last minute before feature freeze.

**Comment by mhvk** [comment]
Thanks, all, merged!

**Pull request merged**
